Modify generateExit to return full set of signing keys

rnicoll opened the pull request:
Modify generateBuild to return full set of signing keys in preparation for anonymity work meaning that owner and issuer keys are typically not the same.

rnicoll commented:
@mikehearn can this be reviewed please? It underpins #714

mikehearn commented on core/src/main/kotlin/net/corda/core/node/ServiceHub.kt:
```diff
@@ -179,6 +180,23 @@ interface ServiceHub : ServicesForResolution {
+     * If the [PublicKey] is actually a [CompositeKey] the first leaf key found locally will be used for signing.
+     * @return The [DigitalSignature.WithKey] generated by signing with the internally held [PrivateKey].
+     */
+    fun createSignature(wireTransaction: WireTransaction, publicKey: PublicKey): DigitalSignature.WithKey = keyManagementService.sign(wireTransaction.id.bytes, publicKey)
```
@roger3cev @mnesbit  Is this compatible with where you want to take the transaction signing APIs? I guessed that ServiceHub lacks createSignature(WireTranscation) currently for a reason?

mnesbit replied:
I agree signing WireTransactions sounds wrong. There is already a method that takes multiple keys to seal a TransactionBuilder.

mikehearn commented on core/src/main/kotlin/net/corda/core/node/ServiceHub.kt:
```diff
@@ -179,6 +180,23 @@ interface ServiceHub : ServicesForResolution {
+    fun createSignature(wireTransaction: WireTransaction, publicKey: PublicKey): DigitalSignature.WithKey = keyManagementService.sign(wireTransaction.id.bytes, publicKey)
+
+    /**
+     * Helper method to create an additional signature for an existing (partially) SignedTransaction
```
This comment doesn't seem to match what it does.

mikehearn commented on finance/src/main/kotlin/net/corda/contracts/asset/Obligation.kt:
```diff
@@ -455,11 +455,11 @@ class Obligation<P : Any> : Contract {
      * @param assetStates the asset states to take funds from. No checks are done about ownership of these states, it is
      * the responsibility of the caller to check that they do not exit funds held by others.
-     * @return the public key of the assets issuer, who must sign the transaction for it to be valid.
+     * @return the public keys who must sign the transaction for it to be valid.
```
who -> which

mikehearn commented on finance/src/main/kotlin/net/corda/contracts/asset/OnLedgerAsset.kt:
```diff
@@ -226,11 +228,11 @@ abstract class OnLedgerAsset<T : Any, C : CommandData, S : FungibleAsset<T>> : C
      * @param assetStates the asset states to take funds from. No checks are done about ownership of these states, it is
      * the responsibility of the caller to check that they do not exit funds held by others.
-     * @return the public key of the assets issuer, who must sign the transaction for it to be valid.
+     * @return the public keys who must sign the transaction for it to be valid.
```
who -> which

mikehearn commented on finance/src/main/kotlin/net/corda/contracts/clause/AbstractConserveAmount.kt:
```diff
@@ -26,15 +26,15 @@ abstract class AbstractConserveAmount<S : FungibleAsset<T>, C : CommandData, T :
      * @param assetStates the asset states to take funds from. No checks are done about ownership of these states, it is
      * the responsibility of the caller to check that they do not attempt to exit funds held by others.
-     * @return the public key of the assets issuer, who must sign the transaction for it to be valid.
+     * @return the public keys who must sign the transaction for it to be valid.
```
who -> which

mikehearn commented on finance/src/main/kotlin/net/corda/flows/CashExitFlow.kt:
```diff
@@ -62,7 +62,9 @@ class CashExitFlow(val amount: Amount<Currency>, val issueRef: OpaqueBytes, prog
         // Sign transaction
         progressTracker.currentStep = SIGNING_TX
-        val tx = serviceHub.signInitialTransaction(builder)
+        val wtx = builder.toWireTransaction()
```
I'm not sure this is the right API. Again, @mnesbit would like your views here?

mnesbit replied:
Yes I would like to seal the builder in one step and I think there is already an API that can do this. We may need to filter out keys that are ours, versus those that are other peoples. However, I would have felt we can determine this from the keys set on the KeyManagementService, or better yet by context/deanonymised Party data. My ideal is that we never use anonymised Party entries in flows and there is a helper for TransactionBuilders to subtitute anonymous keys via Kryo at the point we freeze and sign. Then on message receipt we should be able to use a different Kryo to de-anonymise the Parties. We did discuss this with @mikehearn originaly, and I still think it is a valid way of making this stuff invisible to the flow user.

rnicoll commented:
Removed the functions added to `ServiceHub`, having had it pointed out that there's a `signInitialTransaction` that takes in a list of keys. Modified`signInitialTransaction` to take in any iterable rather than only lists, too.
